modules/xml/test/src/smithy4s/xml/XmlDecodingSpec.scala

```diff
@@ -0,0 +1,318 @@
+
+    val xml = """|<Foo>
+                |   <foos>
+                |      <member>1</member>
+                |      <member>2</member>
+                |      <member>3</member>
+                |   </foos>
+                |</Foo>""".stripMargin
+    testDecode(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("list: custom names") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int.addHints(XmlName("x")))
+          .required[Foo]("foos", _.foos)
+        struct(foos)(Foo.apply)
+      }
+    }
+
+    val xml = """|<Foo>
+                |   <foos>
+                |      <x>1</x>
+                |      <x>2</x>
+                |      <x>3</x>
+                |   </foos>
+                |</Foo>""".stripMargin
+    testDecode(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("list: flattened") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+          .addHints(XmlFlattened())
+        struct(foos)(Foo.apply)
+      }
+    }
+
+    val xml = """|<Foo>
+                 |   <foos>1</foos>
+                 |   <foos>2</foos>
+                 |   <foos>3</foos>
+                 |</Foo>
+                 |""".stripMargin
+    testDecode(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("list: flattened custom names") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+          .addHints(XmlFlattened(), XmlName("x"))
+        struct(foos)(Foo.apply)
+      }
+    }
+
+    val xml = """|<Foo>
+                 |   <x>1</x>
+                 |   <x>2</x>
+                 |   <x>3</x>
+                 |</Foo>
+                 |""".stripMargin
+    testDecode(xml, Foo(List(1, 2, 3)))
+  }
+
+  test("recursion") {
+    case class Foo(foo: Option[Foo])
+    object Foo {
+      implicit val schema: Schema[Foo] = recursive {
+        val foos = schema.optional[Foo]("foo", _.foo)
+        struct(foos)(Foo.apply)
+      }
+    }
+
+    val xml = """|<Foo>
+                 |   <foo>
+                 |      <foo>
+                 |      </foo>
+                 |   </foo>
+                 |</Foo>
+                 |""".stripMargin
+    testDecode(xml, Foo(Some(Foo(Some(Foo(None))))))
+  }
+
+  def testDecode[A: Schema](xml: String, expected: A)(implicit
+      loc: SourceLocation
+  ): IO[Expectations] = {
+    decode[A](xml).map(result => expect.same(result, expected))
+  }
+
+  private def decode[A: Schema](xmlString: String): IO[A] = {
+    val decoder = implicitly[Schema[A]].compile(XmlSchemaVisitor)
+    Stream
+      .emit(xmlString)
+      .through(events[IO, String]())
+      .through(documents[IO, XmlDocument])
+      .take(1)
+      .compile
+      .last
+      .flatMap(_.liftTo[IO](new Throwable("BOOM")))
+      .map(XmlCursor.fromDocument)
```

Comment:
thoughts on adding a `decoder.readDocument` that hides this?

Reply:
Dude the PR is still very much a draft, so keep your guns holstered (I'm merely raising it in case in case I get kicked in the head by a horse, so that the code doesn't die with me), but yeah that's where I was heading towards 😄

Reply:
> my brother in Christ, you asked for the review :trollface:

Nah I said "if you want to have fun", implying that XML is a lot of fun :trollface: